File: assignment 5/c/programs/9.c
#include<stdio.h>
#include<stdlib.h>

// Structure for a node
struct Node {
    int data;
    struct Node *next;
    struct Node *prev;
};

// Function to insert at the end
void insert(struct Node** start, int value)
{
    // If the list is empty, create a single node
    if (*start == NULL) {
        struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
        new_node->data = value;
        new_node->next = new_node->prev = NULL;
        *start = new_node;
    } else {
        struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
        struct Node* last = *start;
        new_node->data = value;
        new_node->next = NULL;
        while (last->next != NULL)
            last = last->next;
        last->next = new_node;
        new_node->prev = last;
    }
}

// Function to print nodes in a given doubly linked list
void printList(struct Node* node)
{
    while (node != NULL) {
        printf("%d ", node->data);
        node = node->next;
    }
}

// Function to reverse a doubly linked list
struct Node* reverse(struct Node* node)
{
    struct Node* temp = NULL;
    struct Node* current = node;

    // Swap next and prev for all nodes of doubly linked list
    while (current != NULL) {
        temp = current->prev;
        current->prev = current->next;
        current->next = temp;
        current = current->prev;
    }

    // Before changing the head, check for the cases like empty list and list with only one node
    if(temp != NULL )
        node = temp->prev;
    return node;
}

// Driver program to test above
int main()
{
    struct Node* head = NULL;
    insert(&head, 1);
    insert(&head, 2);
    insert(&head, 3);
    insert(&head, 4);
    insert(&head, 5);
    insert(&head, 6);

    printf("Original Doubly linked list:\n");
    printList(head);

    head = reverse(head);

    printf("\nDoubly linked list after reversal:\n");
    printList(head);

    return 0;
}

File: assignment 5/python/programs/9.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None

    # Insert a node at the end of the doubly linked list
    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next is not None:
                current = current.next
            current.next = new_node
            new_node.prev = current

    # Reverse the doubly linked list
    def reverse(self):
        if self.head is None or self.head.next is None:
            return

        prev_node = None
        current = self.head

        while current is not None:
            next_node = current.next
            current.next = prev_node
            current.prev = next_node
            prev_node = current
            current = next_node

        self.head = prev_node

    # Display the doubly linked list
    def display(self):
        if self.head is None:
            print("Doubly Linked List is empty.")
            return

        current = self.head
        while current is not None:
            print(current.data, end=" <-> ")
            current = current.next
        print("None")


if __name__ == "__main__":
    doubly_linked_list = DoublyLinkedList()

    # Insert elements into the doubly linked list
    for data in [1, 2, 3, 4, 5]:
        doubly_linked_list.insert_at_end(data)

    # Display the original doubly linked list
    print("Original Doubly Linked List:")
    doubly_linked_list.display()

    # Reverse the doubly linked list
    doubly_linked_list.reverse()

    # Display the reversed doubly linked list
    print("Reversed Doubly Linked List:")
    doubly_linked_list.display()
